CRM/Contribute/Page/Premium.php

```diff
@@ -166,9 +166,9 @@ public function browse() {
           'Premium',
           $dao->id
         );
-        //Financial Type
-        if (!empty($dao->financial_type_id)) {
-          $premiums[$productDAO->id]['financial_type_id'] = CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialType', $dao->financial_type_id, 'name');
+        // Financial Type
+        if (!empty($premiums[$productDAO->id]['financial_type_id'])) {
+          $premiums[$productDAO->id]['financial_type'] = CRM_Core_PseudoConstant::getLabel('CRM_Financial_BAO_FinancialType', 'financial_type', $premiums[$productDAO->id]['financial_type_id']);
```

Comment:
When you add premiums to Contribution page you are allowed to change the Financial type (which is preselected from civicrm_product.financial_type_id). So this should be retrieved from civicrm_premiums_product not from civicrm_product.